## Lab 4: Pandas Continued

**This assignment should be completed by Tuesday February 25, 2020 at 11:59 AM.**

[Pandas](https://pandas.pydata.org/) is one of the most widely used Python libraries in data science. In this lab, you will learn commonly used data wrangling operations/tools in Pandas. We plan to continue the discussion from Lab 3 learning about 

* Grouping dataframes
* Merging dataframes

We will use the baby names dataset from the Social Security Administration stratified by state.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

import os

## Grouping Tables

Load the data from `NY_babynames.csv`

In [2]:
ny = pd.read_csv(os.environ["HOME"] + "/shared/NY_babynames.csv")

### Question 1a: value_counts
To count the number of instances of each unique value in a `Series`, we can use the `value_counts()` [method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.value_counts.html) as `df["col_name"].value_counts()`. 

Count the number of different names for each Year in `NY` (New York).  (You may use the `ny` DataFrame created above.)

**Note:** *We are not computing the number of babies but instead the number of names (rows in the table) for each year.*

In [3]:
num_of_names_per_year = ny["Year"].value_counts()

# YOUR CODE HERE
#raise NotImplementedError()
num_of_names_per_year 

2008    4733
2009    4725
2010    4701
2012    4682
2007    4680
        ... 
1914    1218
1913    1057
1912     996
1911     824
1910     731
Name: Year, Length: 109, dtype: int64

In [4]:
# TEST
assert num_of_names_per_year[2007] == 4680


In [5]:
# TEST
assert num_of_names_per_year[:5].sum() == 23521


### Question 1b
Count the number of different names for each gender in `NY`.



In [6]:
num_of_names_per_gender = ny["Sex"].value_counts()
# YOUR CODE HERE
#raise NotImplementedError()
num_of_names_per_gender

F    169145
M    127054
Name: Sex, dtype: int64

In [7]:
# TEST
assert num_of_names_per_gender["F"] == 169145


### Question 2: groupby ###

Before we jump into using the [`groupby`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) function in Pandas, let's recap how grouping works in general for tabular data through a guided set of questions based on a small dataset of movies and genres. Please see Week 4 lecture slides for a review of grouping.

**Problem Setting:** Below is a dataframe with 5 columns: name of the movie as a `string`, the genre of the movie as a `string`, the first name of the director of the movie as a `string`, the average rating out of 10 on Rotten Tomatoes as an `integer`, and the total gross revenue made by the movie as an `integer`. The point of these guided questions (parts a and b) below is to understand how grouping of data works in general, **not** how grouping works in code. We will worry about how grouping works in Pandas in 7c, which will follow.

Below is the `movies` dataframe we are using, imported from the `movies.csv` file.

In [8]:
movies = pd.read_csv(os.environ["HOME"] + "/shared/movies.csv")

### Question 2a

If we grouped the `movies` dataframe above by `genre`, how many groups would be in the output and what would be the groups? Assign `num_groups` to the number of groups created (hard-code) and fill in `genre_list` as a list containing the names of genres as strings that represent the groups.



In [27]:
print(movies.nunique())
print(movies['genre'].unique())
num_groups = 6
genre_list = ['Action & Adventure', 'Comedy' , 'Science Fiction & Fantasy', 'Drama',
 'Animation', 'Horror']

# YOUR CODE HERE
#raise NotImplementedError()
#genre_list

director    18
genre        6
movie       18
rating       4
revenue     18
dtype: int64
['Action & Adventure' 'Comedy' 'Science Fiction & Fantasy' 'Drama'
 'Animation' 'Horror']


In [28]:
sorted(genre_list)

['Action & Adventure',
 'Animation',
 'Comedy',
 'Drama',
 'Horror',
 'Science Fiction & Fantasy']

In [10]:
# TEST
assert num_groups == 6


### Question 2b

Whenever we group tabular data, it is usually the case that we need to aggregate values from the ungrouped column(s). If we were to group the `movies` dataframe above by `genre`, which column(s) in the `movies` dataframe would it make sense to aggregate if we were interested in finding how well each genre did in the eyes of people? Fill in `agg_cols` with the column name(s).


In [11]:
agg_cols = ['rating', 'revenue']
# YOUR CODE HERE
#raise NotImplementedError()
movies.head()

director                      genre                    movie  rating  \
0     David         Action & Adventure               Deadpool 2       7   
1      Bill                     Comedy                Book Club       5   
2       Ron  Science Fiction & Fantasy  Solo: A Star Wars Story       6   
3  Baltasar                      Drama                   Adrift       6   
4      Bart                      Drama         American Animals       6   

     revenue  
0  318344544  
1   68566296  
2  213476293  
3   31445012  
4    2847319

In [12]:
# TEST
assert sorted(agg_cols) == ['rating', 'revenue']

Now, let's see `groupby` in action, instead of keeping everything abstract. To aggregate data in Pandas, we use the `.groupby()` [function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html). The code below will group the `movies` dataframe by `genre` and find the average revenue and rating for each genre. You can verify you had the same number of groups as what you answered in 2a. 

In [13]:
movies.loc[:, ['genre', 'rating', 'revenue']].groupby('genre').mean()

rating      revenue
genre                                           
Action & Adventure         6.333333  153569934.5
Animation                  5.000000  374408165.0
Comedy                     6.000000   56719237.4
Drama                      6.000000   17146165.5
Horror                     7.000000   68765655.0
Science Fiction & Fantasy  6.000000  312674899.0

### Question 2c

Let's move back to baby names and specifically, the `ny` dataframe. Find the sum of `Count` for each `Name` in the `ny` table. You should use `df.groupby("col_name").sum()`. Your result should be a Pandas Series.

**Note:** *In this question we are now computing the number of registered babies with a given name.*


In [14]:
count_for_names = ny.groupby(['Name'])['Count'].sum()
# YOUR CODE HERE
#raise NotImplementedError()

count_for_names
count_for_names.sort_values(ascending=False)[:5]

Name
John       497670
Robert     440880
Michael    438074
James      344839
Joseph     344219
Name: Count, dtype: int64

In [15]:
# TEST
assert count_for_names["Michael"] == 438074


In [16]:
# TEST
assert count_for_names[:100].sum() == 72157


### Question 2d

Find the sum of `Count` for each female name after year 1999 (`>1999`) in New York.


In [17]:
female_name_count = ny[(ny['Sex'] == 'F') & (ny['Year'] > 1999)].groupby(['Name'])['Count'].sum()
# YOUR CODE HERE
#raise NotImplementedError()


female_name_count.sort_values(ascending=False)[:5]

Name
Emily       21485
Isabella    21446
Olivia      21420
Emma        20587
Sophia      19883
Name: Count, dtype: int64

In [18]:
# TEST
assert female_name_count["Emily"] == 21485


In [19]:
# TEST
assert female_name_count["Isabella"] == 21446


### Question 3: Grouping Multiple Columns

Let's move back to the `movies` dataframe. Which of the following lines of code will output the following dataframe? Write your answer (hard-coded) as either 1, 2, 3, or 4. Recall that the arguments to `pd.pivot_table` are as follows: `data` is the input dataframe, `index` includes the values we use as rows, `columns` are the columns of the pivot table, `values` are the values in the pivot table, and `aggfunc` is the aggregation function that we use to aggregate `values`.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>rating</th>
      <th>5</th>
      <th>6</th>
      <th>7</th>
      <th>8</th>
    </tr>
    <tr>
      <th>genre</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Action &amp; Adventure</th>
      <td>208681866.0</td>
      <td>129228350.0</td>
      <td>318344544.0</td>
      <td>6708147.0</td>
    </tr>
    <tr>
      <th>Animation</th>
      <td>374408165.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Comedy</th>
      <td>55383976.0</td>
      <td>30561590.0</td>
      <td>NaN</td>
      <td>111705055.0</td>
    </tr>
    <tr>
      <th>Drama</th>
      <td>NaN</td>
      <td>17146165.5</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Horror</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>68765655.0</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Science Fiction &amp; Fantasy</th>
      <td>NaN</td>
      <td>312674899.0</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

1) `pd.pivot_table(data=movies, index='genre', columns='rating', values='revenue', aggfunc=np.mean)`

2) `movies.groupby(['genre', 'rating'])['revenue'].mean()`

3) `pd.pivot_table(data=movies, index='rating', columns='genre', values='revenue', aggfunc=np.mean)`

4) `movies.groupby('revenue')[['genre', 'rating']].mean()`


In [20]:
q3_answer = 1
# YOUR CODE HERE
#raise NotImplementedError()

### Question 4: Merging


#### Question 4a

Time to put everything together! Merge `movies` and `count_for_names` to find the number of registered baby names for each director using [`pd.merge`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html?highlight=merge#pandas.merge). Only include names that appear in both `movies` and `count_for_names`.

**Hint:** You might need to convert the `count_for_names` series to a dataframe. Take a look at the ``to_frame`` method of a series to do this. 

Your first row should look something like this:

**Note**: It is ok if you have 2 separate columns with names instead of just one column.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>director</th>
      <th>genre</th>
      <th>movie</th>
      <th>rating</th>
      <th>revenue</th>
      <th>Count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>David</td>
      <td>Action &amp; Adventure</td>
      <td>Deadpool 2</td>
      <td>7</td>
      <td>318344544</td>
      <td>371646</td>
    </tr>
  </tbody>
</table>
</table>


In [21]:

merged_df =  movies.merge(count_for_names, how = 'inner', left_on = 'director', right_on = 'Name', right_index = True)
# YOUR CODE HERE
#raise NotImplementedError()

merged_df

director                      genre                        movie  \
0         David         Action & Adventure                   Deadpool 2   
1          Bill                     Comedy                    Book Club   
2           Ron  Science Fiction & Fantasy      Solo: A Star Wars Story   
4          Bart                      Drama             American Animals   
5          Gary         Action & Adventure                     Oceans 8   
6          Drew         Action & Adventure                Hotel Artemis   
7          Brad                  Animation                Incredibles 2   
8          Jeff                     Comedy                          Tag   
10      Charles                     Comedy                   Uncle Drew   
11       Gerard                     Horror              The First Purge   
12       Peyton         Action & Adventure         Ant-Man and the Wasp   
16  Christopher         Action & Adventure  Mission: Impossible-Fallout   
17         Marc                     Comedy           Christopher Robbin   

    rating    revenue   Count  
0        7  318344544  280681  
1        5   68566296    3949  
2        6  213476293    1367  
4        6    2847319    1145  
5        6  138803463   62840  
6        8    6708147    4192  
7        5  594398019    3407  
8        6   54336863    4707  
10       5   42201656  161375  
11       7   68765655   16924  
12       5  208681866    4036  
16       6  182080372  171144  
17       6    6786317   21322

In [22]:
# TEST
assert merged_df.loc[0, 'Count'] == 280681



In [23]:
# TEST
assert merged_df.loc[7, 'Count'] == 3407


In [24]:
# TEST
assert len(merged_df) == 13

#### Question 4b

How many directors in the original `movies` table did not get included in the `merged_df` dataframe? Please hard-code your answer as a number in `q4b`, then explain your answer in 1-2 sentences as a comment below.



In [25]:
merged_df_left = movies.merge(count_for_names, how = 'left', left_on = 'director', right_on = 'Name')
print(merged_df_left.head())
print()
print(pd.isnull(merged_df_left["Count"]).sum())
q_4b = 5

# YOUR CODE HERE
#raise NotImplementedError()

# Explain your solution:
'''I used the left join to include the all of the directors from the movies dataframe even if they
did not appear in the count_for_names series. If the director's name did not have a value to join with 
in the count_for_names series a null value would appear in its count column, since that is what is being 
referenced in join. I simply used the isnull function create a boolean output series where the true would 
equal a null value in the count column. I then summed the trues to give me a count for the missing directors.
'''

   director                      genre                    movie  rating  \
0     David         Action & Adventure               Deadpool 2       7   
1      Bill                     Comedy                Book Club       5   
2       Ron  Science Fiction & Fantasy  Solo: A Star Wars Story       6   
3  Baltasar                      Drama                   Adrift       6   
4      Bart                      Drama         American Animals       6   

     revenue     Count  
0  318344544  280681.0  
1   68566296    3949.0  
2  213476293    1367.0  
3   31445012       NaN  
4    2847319    1145.0  

5


"I used the left join to include the all of the directors from the movies dataframe even if they\ndid not appear in the count_for_names series. If the director's name did not have a value to join with \nin the count_for_names series a null value would appear in its count column, since that is what is being \nreferenced in join. I simply used the isnull function create a boolean output series where the true would \nequal a null value in the count column. I then summed the trues to give me a count for the missing directors.\n"